lib/grizzly/zwave_firmware.ex

```diff
@@ -0,0 +1,115 @@
+defmodule Grizzly.ZwaveFirmware do
+  @moduledoc false
+
+  require Logger
+
+  @doc """
+  Update the ZWave chip firmware if zw_programmer is present and a
+  valid GBL file is available that would up the firmware version.
+  """
+  @spec maybe_run_zwave_firmware_update(Grizzly.Options.t()) :: :ok
+  def maybe_run_zwave_firmware_update(options) do
+    with :ok <- check_zwave_programmer_exists(options),
+         {:ok, gbl_file} <- get_firmware_file(options),
+         :ok <- maybe_upload_zwave_firmware_update(gbl_file, options) do
+      :ok
+    else
+      {:error, reason} ->
+        Logger.info("[Grizzly] The hub's z-wave firmware was NOT updated: #{inspect(reason)}")
+    end
+
+    :ok
+  end
+
+  defp check_zwave_programmer_exists(options) do
+    if File.exists?(options.zw_programmer_path) do
+      :ok
+    else
+      {:error, :zw_programmer_does_not_exist}
+    end
+  end
+
+  defp get_firmware_file(options) do
+    file = options.zwave_firmware_path
+
+    if file != nil and File.exists?(file) do
+      {:ok, file}
+    else
+      {:error, :no_zwave_firmware_file}
+    end
+  end
+
+  defp maybe_upload_zwave_firmware_update(gbl_file, options) do
+    with {:ok, gbl_version} <- gbl_version(gbl_file),
+         {:ok, current_version} <- current_version(options) do
+      case validate_upgrade(current_version, gbl_version) do
+        :ok ->
+          upload_zwave_firmware_update(gbl_file, options)
+
+        {:error, :not_an_upgrade} ->
+          Logger.info("[Grizzly] The hub's zwave firmware is already up-to-date")
+          :ok
+      end
+    end
+  end
+
+  # .../ZW_SerialAPI_Controller_7.16.3_297_ZGM130S_REGION_US_LR.gbl
+  defp gbl_version(gbl_file) do
+    basename = Path.basename(gbl_file)
+    regex = ~r/\w+_(?<major>\d+)\.(?<minor>\d+)\.(?<patch>\d+).*\.gbl/
```

Comment:
This seems like it might be fragile. I'm leaning to manually putting firmware versions in the `zwave_firmware` map to avoid worrying about upstream file naming conventions.